packages/core-data/src/test/reducer.js

```diff
@@ -99,6 +107,206 @@ describe( 'entities', () => {
 	} );
 } );
 
+describe( 'undo', () => {
+	let lastEdits;
+	let undoState;
+	let expectedUndoState;
+	const createEditActionPart = ( edits ) => ( {
+		kind: 'someKind',
+		name: 'someName',
+		recordId: 'someRecordId',
+		edits,
+	} );
+	const createNextEditAction = ( edits, transientEdits = {} ) => {
+		let action = {
+			...createEditActionPart( edits ),
+			transientEdits,
+		};
+		action = {
+			type: 'EDIT_ENTITY_RECORD',
+			...action,
+			meta: {
+				undo: { ...action, edits: lastEdits },
+			},
+		};
+		lastEdits = { ...lastEdits, ...edits };
+		return action;
+	};
+	const createNextUndoState = ( ...args ) => {
+		let action = {};
+		if ( args[ 0 ] === 'isUndo' || args[ 0 ] === 'isRedo' ) {
+			// We need to "apply" the undo level here and build
+			// the action to move the offset.
```

Comment:
Why does the action need to be build based on the state? Isn't that something that can be done in the reducer?

Reply:
Previously we had `UNDO` and `REDO` actions.

Reply:
Because the editor grabs the undo edit and applies it, the undo reducer just keeps track of the offset.